kvdb/postgres/readwrite_tx.go

```diff
@@ -0,0 +1,386 @@
+// +build kvdb_postgres
+
+package postgres
+
+import (
+	"bytes"
+	"database/sql"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/btcsuite/btcwallet/walletdb"
+)
+
+// readWriteTx holds a reference to an open postgres transaction.
+type readWriteTx struct {
+	db *db
+	tx *sql.Tx
+
+	// onCommit gets called upon commit.
+	onCommit func()
+
+	// active is true if the transaction hasn't been committed yet.
+	active bool
+
+	// locker is a pointer to the global db lock.
+	locker sync.Locker
+}
+
+// newReadWriteTx creates an rw transaction using a connection from the
+// specified pool.
+func newReadWriteTx(db *db, readOnly bool) (*readWriteTx, error) {
+	// Obtain the global lock instance. An alternative here is to obtain a
+	// database lock from Postgres. Unfortunately there is no database-level
+	// lock in Postgres, meaning that each table would need to be locked
+	// individually. Perhaps an advisory lock could perform this function
+	// too.
+	var locker sync.Locker = &db.lock
+	if readOnly {
+		locker = db.lock.RLocker()
+	}
+	locker.Lock()
+
+	tx, err := db.db.Begin()
+	if err != nil {
+		locker.Unlock()
+		return nil, err
+	}
+
+	return &readWriteTx{
+		db:     db,
+		tx:     tx,
+		active: true,
+		locker: locker,
+	}, nil
+}
+
+// hasSpecialChars returns true if any of the characters in the given string
+// cannot be used in postgres table names.
+func hasSpecialChars(s string) bool {
+	for _, b := range s {
+		if !(b >= 'a' && b <= 'z') &&
+			!(b >= '0' && b <= '9') && b != '-' {
+
+			return true
+		}
+	}
+
+	return false
+}
+
+// toTableName attempts to convert a binary key to a sql-compatible table name.
+// If the mapping can't be made, an empty string is returned.
+func (tx *readWriteTx) toTableName(key []byte) string {
+	// Max table name length in postgres is 63, but keep some slack for
+	// index postfixes.
+	const maxTableNameLen = 50
+
+	var table string
+	if hasSpecialChars(string(key)) {
+		return ""
+	}
+
+	table = strings.Replace(string(key), "-", "_", -1)
+	table = tx.db.getPrefixedTableName(table)
+
+	if len(table) > maxTableNameLen {
+		return ""
+	}
+
+	return table
+}
+
+// ReadBucket opens the root bucket for read only access.  If the bucket
+// described by the key does not exist, nil is returned.
+func (tx *readWriteTx) ReadBucket(key []byte) walletdb.ReadBucket {
+	bucket := tx.ReadWriteBucket(key)
```

Comment:
nit: If I'm not mistaken you can just `return tx.ReadWriteBucket(key)`

Reply:
fixed